distributor.go

```diff
@@ -176,53 +184,87 @@ func (d *Distributor) Append(ctx context.Context, samples []*model.Sample) error
 
 	d.receivedSamples.Add(float64(len(samples)))
 
-	samplesByIngester := map[string][]*model.Sample{}
-	for _, sample := range samples {
-		key := tokenForMetric(userID, sample.Metric)
-		ingesters, err := d.cfg.Ring.Get(key, d.cfg.ReplicationFactor)
-		if err != nil {
-			return err
+	sampleTrackers := make([]sampleTracker, len(samples), len(samples))
+	samplesByIngester := map[string][]*sampleTracker{}
+	keys := make([]uint32, len(samples), len(samples))
+
+	for i, sample := range samples {
+		keys[i] = tokenForMetric(userID, sample.Metric)
+	}
+
+	ingesters, err := d.cfg.Ring.BatchGet(keys, d.cfg.ReplicationFactor, ring.Write)
+	if err != nil {
+		return err
+	}
+
+	for i := range samples {
+		sampleTrackers[i] = sampleTracker{
+			sample:     samples[i],
+			minSuccess: (len(ingesters[i]) / 2) + 1,
+			succeeded:  0,
+		}
+
+		// Skip those that have not heartbeated in a while. NB these are still
+		// included in the calculation of minSuccess, so if too many failed ingesters
+		// will cause the whole write to fail.
```

Comment:
Ah; the check its referring to is actually at the bottom of the function (`minSuccess` is compared to `succeeded`).  The check just below the commend is just a shortcut - if there are not minSuccess available ingesters, don't even both trying.

Will add this as a comment.